### ISSUE TITLE
Why BoundServiceAccountTokenVolume featureGate is not working?

### ISSUE THREAD

**Issue by a8j8i8t8** [kubernetes/kops#8451]
**1. What `kops` version are you running? The command `kops version`, will display
 this information.**
1.15.1
**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**
1.15.9
**3. What cloud provider are you using?**
aws
**4. What commands did you run?  What is the simplest way to reproduce this issue?**
Created cluster with below feature gate enabled for api-server.
```
kubeAPIServer:
    featureGates:
        BoundServiceAccountTokenVolume: true
```
**5. What happened after the commands executed?**
kube-api-server pod didn't start on master nodes.
**6. Anything else do we need to know?**
Is any other configuration needed for this feature gate to work properly?

**Comment by robinpercy** [comment]
Thanks @a8j8i8t8 . I was able to reproduce the issue based on your description. Need to investigate further.  Will let you know if I find additional required configs.

**Comment by robinpercy** [comment]
@a8j8i8t8 looks like you also need to specify the `ServiceAccountIssuer` and `ServiceAccountSigningKeyFile`

Here's the error reported by the api-server
```service-account-issuer is a required flag when BoundServiceAccountTokenVolume is enabled, --service-account-signing-key-file and --service-account-issuer are required flags```

**Comment by rvirdiz** [comment]
@robinpercy were you able to solve this issue. It tried by specifying the required values, but no luck.
--service-account-issuer
--service-account-signing-key-file
--service-account-key-file
--service-account-api-audiences (--api-audiences now)

I am using kops 1.15.2 on K8s cluster 1.15.10 on AWS, to install istio-1.3.2, which requires the above values to be added to kube-apiserver, I am not sure what values needs to be setup for kops managed cluster, please let me know.
https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-token-volume-projection

**Comment by mathieu-gilloots** [comment]
Same problem here, with istio sds, we need to add the extra flags like here : 
https://github.com/istio/istio/blob/master/prow/config/trustworthy-jwt.yaml
How to process to do that with kops ?
Thanks

**Comment by rvirdiz** [comment]
> Same problem here, with istio sds, we need to add the extra flags like here :
> https://github.com/istio/istio/blob/master/prow/config/trustworthy-jwt.yaml
> How to process to do that with kops ?
> Thanks

I tried to made changes in K8s 1.15 API server manifest, and restarted kubelet service, it worked for me.

/etc/kubernetes/manifests/kube-apiserver.manifest, added following parameters to this file.
    - --feature-gates=TokenRequest=true
    - --service-account-issuer=kubernetes.default.svc
    - --service-account-signing-key-file=/srv/kubernetes/server.key
    - --service-account-key-file=/srv/kubernetes/server.key
    - --api-audiences=api,istio-ca

**Comment by rvirdiz** [comment]
For kops, add following parameters to your cluster spec yaml file.
spec:
  kubeAPIServer:
    apiAudiences:
    - api
    - istio-ca
    featureGates:
      TokenRequest: "true"
    serviceAccountIssuer: kubernetes.default.svc
    serviceAccountKeyFile:
    - /srv/kubernetes/server.key
    serviceAccountSigningKeyFile: /srv/kubernetes/server.key

**Comment by jmymy** [comment]
working on updating docs https://github.com/istio/istio.io/pull/6744

> For kops, add following parameters to your cluster spec yaml file to deploy istio-1.3+ SDS by enabling ServiceAccountTokenVolumeProjection feature.
> https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-token-volume-projection
> 
> ```
> spec:
>   kubeAPIServer:
>     apiAudiences:
>     - api
>     - istio-ca
>     featureGates:
>       TokenRequest: "true"
>     serviceAccountIssuer: kubernetes.default.svc
>     serviceAccountKeyFile:
>     - /srv/kubernetes/server.key
>     serviceAccountSigningKeyFile: /srv/kubernetes/server.key
> ```

note, featureGates for TokenRequest is not needed in k8s 1.12+

**Comment by olemarkus** [comment]
This one can be closed now?

**Comment by adammw** [comment]
`/srv/kubernetes/server.key` seems to be generated unique to each server on the most recent version of kops and I believe this configuration can actually cause token projection to not work correctly with multiple API servers. In contrast, leaving out the `serviceAccountKeyFile` and `serviceAccountSigningKeyFile`, kops will instead configure to use `/srv/kubernetes/service-account.key` which is stable between API servers.

**Comment by johngmyers** [comment]
@adammw that is correct: one should not override either `serviceAccountKeyFile` or `serviceAccountSigningKeyFile` in kops 1.19+.

**Comment by adammw** [comment]
aha - found the relevant PRs: https://github.com/kubernetes/kops/pull/10712 and https://github.com/kubernetes/kops/pull/9534 - although I couldn't see any specific call out in the 1.19 CHANGELOG so we ran over this issue upgrading from kops 1.18
